Support for elecgator EtherCATduino board

pieterjanbuntinx opened the pull request:
**Summary**

This PR adds support for the [elecgator ](https://www.elecgator.com/) EtherCATduino board. This board is based on a STM32F746ZET microcontroller and uses a LAN9252 chip as the 2-port EtherCAT device controller (ESC). It has a built-in DC-DC that supports voltages from 7 to 28 Volts, a MicroUSB port, 2 RJ45 connectors, and the same pinout and overall form factor as an Arduino Mega. We are also in the process of creating a library for using the EtherCAT supported on this board. 

This PR fixes/implements the following **features**

* [x] Added support for the elecgator EtherCATduino

**Validation**

* Passed the CheckVariant example sketch
* Successfully builds and uploads using DFU and StLink (SWD)

fpistm commented:
Hi @pieterjanbuntinx 
Thanks for this PR.
Do you have some schematics? user manual ? 
On the official site except the picture of the boards it seems there is nothing.

fpistm commented on boards.txt:
```diff
@@ -5506,6 +5506,40 @@ Midatronics.menu.upload_method.dfuMethod.upload.protocol=2
 Midatronics.menu.upload_method.dfuMethod.upload.options=-g
 Midatronics.menu.upload_method.dfuMethod.upload.tool=stm32CubeProg
 
+################################################################################
+# elecgator boards
+
+elecgator.name=elecgator boards
+
+elecgator.build.core=arduino
+elecgator.build.board=elecgator
+elecgator.build.variant_h=variant_{build.board}.h
+
+# EtherCATduino board
+elecgator.menu.pnum.ETHERCAT_DUINO=EtherCATduino
+elecgator.menu.pnum.ETHERCAT_DUINO.node=NODE_F746ZET
+elecgator.menu.pnum.ETHERCAT_DUINO.upload.maximum_size=524288
+elecgator.menu.pnum.ETHERCAT_DUINO.upload.maximum_data_size=327680
+elecgator.menu.pnum.ETHERCAT_DUINO.build.mcu=cortex-m7
+elecgator.menu.pnum.ETHERCAT_DUINO.build.flags.fp=-mfpu=fpv4-sp-d16 -mfloat-abi=hard
+elecgator.menu.pnum.ETHERCAT_DUINO.build.board=ETHERCAT_DUINO
+elecgator.menu.pnum.ETHERCAT_DUINO.build.series=STM32F7xx
+elecgator.menu.pnum.ETHERCAT_DUINO.build.product_line=STM32F746xx
+elecgator.menu.pnum.ETHERCAT_DUINO.build.variant=STM32F7xx/F745Z(E-G)T_F746Z(E-G)(T-Y)_F750Z8T_F756ZG(T-Y)
+elecgator.menu.pnum.ETHERCAT_DUINO.build.extra_flags=-D{build.product_line} {build.enable_usb} {build.xSerial} {build.bootloader_flags}
+elecgator.menu.pnum.ETHERCAT_DUINO.build.cmsis_lib_gcc=arm_cortexM7lfsp_math
+
+# Upload menu
+elecgator.menu.upload_method.swdMethod=STM32CubeProgrammer (SWD)
+elecgator.menu.upload_method.swdMethod.upload.protocol=0
+elecgator.menu.upload_method.swdMethod.upload.options=-g
+elecgator.menu.upload_method.swdMethod.upload.tool=stm32CubeProg
+
+elecgator.menu.upload_method.dfuMethod=STM32CubeProgrammer (DFU)
+elecgator.menu.upload_method.dfuMethod.upload.protocol=2
+elecgator.menu.upload_method.dfuMethod.upload.options=-g
+elecgator.menu.upload_method.dfuMethod.upload.tool=stm32CubeProg
+
```
Please move it to ESC boards section, as it seems to be an ESC board. we would not add a dedicated section.

fpistm commented on README.md:
```diff
@@ -460,6 +460,12 @@ User can add a STM32 based board following this [wiki](https://github.com/stm32d
 | :----: | :-------: | ---- | :-----: | :---- |
 | :green_heart:  | STM32WB55CE | [SharkyMKR](https://midatronics.com/shop/development-boards/mkr-sharky-i/) | *1.7.0* |  |
 
+### [elecgator](https://www.elecgator.com/) boards
+
+| Status | Device(s) | Name | Release | Notes |
+| :----: | :-------: | ---- | :-----: | :---- |
+| :green_heart:  | STM32F746ZET | [EtherCATduino](https://www.elecgator.com/products/info/ethercatduino) | *2.1.0* |  |
+
```
Please move it to ESC boards section, as it seems to be an ESC board. we would not add a dedicated section.

fpistm commented on boards.txt:
```diff
@@ -5672,6 +5706,12 @@ Midatronics.menu.xserial.none.build.xSerial=-DHAL_UART_MODULE_ENABLED -DHWSERIAL
 Midatronics.menu.xserial.disabled=Disabled (No Serial support)
 Midatronics.menu.xserial.disabled.build.xSerial=
 
+elecgator.menu.xserial.generic=Enabled (generic 'Serial')
+elecgator.menu.xserial.none=Enabled (no generic 'Serial')
+elecgator.menu.xserial.none.build.xSerial=-DHAL_UART_MODULE_ENABLED -DHWSERIAL_NONE
+elecgator.menu.xserial.disabled=Disabled (no Serial support)
+elecgator.menu.xserial.disabled.build.xSerial=
+
```
To be removed when moved to ESC section

fpistm commented on boards.txt:
```diff
@@ -5943,6 +5983,19 @@ Midatronics.menu.xusb.HS.build.usb_speed=-DUSE_USB_HS
 Midatronics.menu.xusb.HSFS=High Speed in Full Speed mode
 Midatronics.menu.xusb.HSFS.build.usb_speed=-DUSE_USB_HS -DUSE_USB_HS_IN_FS
 
+elecgator.menu.usb.none=None
+elecgator.menu.usb.CDCgen=CDC (generic 'Serial' supersede U(S)ART)
+elecgator.menu.usb.CDCgen.build.enable_usb={build.usb_flags} -DUSBD_USE_CDC
+elecgator.menu.usb.CDC=CDC (no generic 'Serial')
+elecgator.menu.usb.CDC.build.enable_usb={build.usb_flags} -DUSBD_USE_CDC -DDISABLE_GENERIC_SERIALUSB
+elecgator.menu.usb.HID=HID (keyboard and mouse)
+elecgator.menu.usb.HID.build.enable_usb={build.usb_flags} -DUSBD_USE_HID_COMPOSITE
+elecgator.menu.xusb.FS=Low/Full Speed
+elecgator.menu.xusb.HS=High Speed
+elecgator.menu.xusb.HS.build.usb_speed=-DUSE_USB_HS
+elecgator.menu.xusb.HSFS=High Speed in Full Speed mode
+elecgator.menu.xusb.HSFS.build.usb_speed=-DUSE_USB_HS -DUSE_USB_HS_IN_FS
+
```
To be removed when moved to ESC section

fpistm commented on boards.txt:
```diff
@@ -6431,6 +6484,25 @@ Midatronics.menu.opt.o3lto.build.flags.optimize=-O3 -flto
 Midatronics.menu.opt.ogstd=Debug (-g)
 Midatronics.menu.opt.ogstd.build.flags.optimize=-g -Og
 
+elecgator.menu.opt.osstd=Smallest (-Os default)
+elecgator.menu.opt.osstd.build.flags.optimize=-Os
+elecgator.menu.opt.oslto=Smallest (-Os) with LTO
+elecgator.menu.opt.oslto.build.flags.optimize=-Os -flto
+elecgator.menu.opt.o1std=Fast (-O1)
+elecgator.menu.opt.o1std.build.flags.optimize=-O1
+elecgator.menu.opt.o1lto=Fast (-O1) with LTO
+elecgator.menu.opt.o1lto.build.flags.optimize=-O1 -flto
+elecgator.menu.opt.o2std=Faster (-O2)
+elecgator.menu.opt.o2std.build.flags.optimize=-O2
+elecgator.menu.opt.o2lto=Faster (-O2) with LTO
+elecgator.menu.opt.o2lto.build.flags.optimize=-O2 -flto
+elecgator.menu.opt.o3std=Fastest (-O3)
+elecgator.menu.opt.o3std.build.flags.optimize=-O3
+elecgator.menu.opt.o3lto=Fastest (-O3) with LTO
+elecgator.menu.opt.o3lto.build.flags.optimize=-O3 -flto
+elecgator.menu.opt.ogstd=Debug (-g)
+elecgator.menu.opt.ogstd.build.flags.optimize=-g -Og
+
```
To be removed when moved to ESC section

fpistm commented on boards.txt:
```diff
@@ -6691,3 +6763,13 @@ Midatronics.menu.rtlib.nanofps=Newlib Nano + Float Printf/Scanf
 Midatronics.menu.rtlib.nanofps.build.flags.ldspecs=--specs=nano.specs -u _printf_float -u _scanf_float
 Midatronics.menu.rtlib.full=Newlib Standard
 Midatronics.menu.rtlib.full.build.flags.ldspecs=
+
+elecgator.menu.rtlib.nano=Newlib Nano (default)
+elecgator.menu.rtlib.nanofp=Newlib Nano + Float Printf
+elecgator.menu.rtlib.nanofp.build.flags.ldspecs=--specs=nano.specs -u _printf_float
+elecgator.menu.rtlib.nanofs=Newlib Nano + Float Scanf
+elecgator.menu.rtlib.nanofs.build.flags.ldspecs=--specs=nano.specs -u _scanf_float
+elecgator.menu.rtlib.nanofps=Newlib Nano + Float Printf/Scanf
+elecgator.menu.rtlib.nanofps.build.flags.ldspecs=--specs=nano.specs -u _printf_float -u _scanf_float
+elecgator.menu.rtlib.full=Newlib Standard
+elecgator.menu.rtlib.full.build.flags.ldspecs=
```
To be removed when moved to ESC section

fpistm commented on variants/STM32F7xx/F745Z(E-G)T_F746Z(E-G)(T-Y)_F750Z8T_F756ZG(T-Y)/variant_ETHERCAT_DUINO.cpp:
```diff
@@ -0,0 +1,152 @@
+  */
+  WEAK void SystemClock_Config(void)
+  {
+    RCC_OscInitTypeDef RCC_OscInitStruct = {0};
+    RCC_ClkInitTypeDef RCC_ClkInitStruct = {0};
+    RCC_PeriphCLKInitTypeDef PeriphClkInitStruct = {0};
```
Remove 0 to avoid warning during the build.

```suggestion
    RCC_OscInitTypeDef RCC_OscInitStruct = {};
    RCC_ClkInitTypeDef RCC_ClkInitStruct = {};
    RCC_PeriphCLKInitTypeDef PeriphClkInitStruct = {};
```

fpistm commented on variants/STM32F7xx/F745Z(E-G)T_F746Z(E-G)(T-Y)_F750Z8T_F756ZG(T-Y)/variant_ETHERCAT_DUINO.cpp:
```diff
@@ -0,0 +1,152 @@
+    RCC_OscInitStruct.PLL.PLLN = 216;
+    RCC_OscInitStruct.PLL.PLLP = RCC_PLLP_DIV2;
+    RCC_OscInitStruct.PLL.PLLQ = 9;
+    if (HAL_RCC_OscConfig(&RCC_OscInitStruct) != HAL_OK)
+    {
+    }
```
Fixiing Astyle issue and add the error handler call.

```suggestion
    if (HAL_RCC_OscConfig(&RCC_OscInitStruct) != HAL_OK)  {
      Error_Handler();
    }
```

fpistm commented on variants/STM32F7xx/F745Z(E-G)T_F746Z(E-G)(T-Y)_F750Z8T_F756ZG(T-Y)/variant_ETHERCAT_DUINO.cpp:
```diff
@@ -0,0 +1,152 @@
+    }
+    /** Activate the Over-Drive mode 
+  */
+    if (HAL_PWREx_EnableOverDrive() != HAL_OK)
+    {
+    }
```
```suggestion
    if (HAL_PWREx_EnableOverDrive() != HAL_OK) {
      Error_Handler();
    }
```

fpistm commented on variants/STM32F7xx/F745Z(E-G)T_F746Z(E-G)(T-Y)_F750Z8T_F756ZG(T-Y)/variant_ETHERCAT_DUINO.cpp:
```diff
@@ -0,0 +1,152 @@
+    RCC_ClkInitStruct.APB1CLKDivider = RCC_HCLK_DIV4;
+    RCC_ClkInitStruct.APB2CLKDivider = RCC_HCLK_DIV2;
+
+    if (HAL_RCC_ClockConfig(&RCC_ClkInitStruct, FLASH_LATENCY_7) != HAL_OK)
+    {
+    }
```
```suggestion
    if (HAL_RCC_ClockConfig(&RCC_ClkInitStruct, FLASH_LATENCY_7) != HAL_OK) {
      Error_Handler();
    }
```

fpistm commented on variants/STM32F7xx/F745Z(E-G)T_F746Z(E-G)(T-Y)_F750Z8T_F756ZG(T-Y)/variant_ETHERCAT_DUINO.cpp:
```diff
@@ -0,0 +1,152 @@
+    PeriphClkInitStruct.Uart8ClockSelection = RCC_UART8CLKSOURCE_PCLK1;
+    PeriphClkInitStruct.I2c1ClockSelection = RCC_I2C1CLKSOURCE_PCLK1;
+    PeriphClkInitStruct.Clk48ClockSelection = RCC_CLK48SOURCE_PLL;
+    if (HAL_RCCEx_PeriphCLKConfig(&PeriphClkInitStruct) != HAL_OK)
+    {
+    }
```
```suggestion
    if (HAL_RCCEx_PeriphCLKConfig(&PeriphClkInitStruct) != HAL_OK) {
      Error_Handler();
    }
```

fpistm commented on variants/STM32F7xx/F745Z(E-G)T_F746Z(E-G)(T-Y)_F750Z8T_F756ZG(T-Y)/variant_ETHERCAT_DUINO.cpp:
```diff
@@ -0,0 +1,152 @@
+    , 6  // A8 
+    , 7  // A9 
+    , 36 // A10
+    , 37 // A11
```
You defined only 12 Analog pins while in the variant.h 
` #define NUM_ANALOG_INPUTS       24`

The check variant should have result failed.

pieterjanbuntinx replied:
With 
```
#define NUM_DIGITAL_PINS             114 OR 54
#define NUM_ANALOG_INPUTS       24
```
CheckVariant compiles successfully. 
With `#define NUM_ANALOG_INPUTS       12` it fails. That is why I left both at the default for that chip, 114 and 24. Any idea why it fails?

fpistm replied:
> CheckVariant compiles successfully.

The goal is not only to compile but to execute it to have the verdict: Passed  or Failed.

pieterjanbuntinx replied:
Okay, that's maybe a stupid thing of me to not do haha, sorry for that.

If I compile with `NUM_ANALOG_INPUTS` defined as 12 I get the following error:
```
 52 | #define PC2                     PIN_A12
      |                                 ^~~~~~~
c:\Users\piete\Downloads\CheckVariant\digitalPinsTest.ino:2:27: note: in expansion of macro 'PC2'
    2 | #define PNUM(Port, Pin)   P ## Port ## Pin
      |                           ^
c:\Users\piete\Downloads\CheckVariant\digitalPinsTest.ino:4:105: note: in expansion of macro 'PNUM'
    4 | #define DECLTESTPNUM(Port, Pin)  checkPinNumber(XSTR(PNAME(Port, Pin)), PNAME(Port, Pin), "P"#Port#Pin, PNUM(Port, Pin))
      |                                                                                                         ^~~~
c:\Users\piete\Downloads\CheckVariant\digitalPinsTest.ino:147:8: note: in expansion of macro 'DECLTESTPNUM'
  147 |   if (!DECLTESTPNUM(C, 2)) testPassed = false;
```

When going to the definition of PIN_A12 it leads me to pins_arduino_analog.h which contains the following:
```
#if NUM_ANALOG_INPUTS > 12
  #define PIN_A12      (PIN_A11 + 1)
#endif
#if NUM_ANALOG_INPUTS > 13
  #define PIN_A13      (PIN_A12 + 1)
#endif
#if NUM_ANALOG_INPUTS > 14
  #define PIN_A14      (PIN_A13 + 1)
#endif
```

So it's fairly clear why the compiler fails. But I don't really see how I can solve this problem. I can change the `NUM_ANALOG_INPUTS` to 24 and declare all the pins in the analog input array but then the pin names would not correspond to the names printed on the silkscreen of our board. Do you have an idea how to solve this?

fpistm replied:
You have to remove all `PIN_Ax `> 11 in the variant.h because they are defined based on the `NUM_ANALOG_INPUTS`
Your variant define which pins have Analog capacity. So if you provide only 12 analog then only PIN_A0 to PIN_A11 have to be define.

pieterjanbuntinx replied:
It was kind of a struggle but I have managed to make it all work: 
```
12:7:38.734 -> #####
12:7:38.737 -> NUM_DIGITAL_PINS = 62
12:7:38.737 -> NUM_ANALOG_INPUTS = 12
12:7:38.737 -> #####
12:7:38.737 -> Checking digital pins...
12:7:38.738 -> End check digital pins
12:7:38.738 -> #####
12:7:38.738 -> Checking analog pins definition...
12:7:38.749 -> End check analog pins
12:7:38.749 -> #####
12:7:38.749 -> Checking IP instances (I2C, UART, SPI)...
12:7:38.768 -> End check IP instances (I2C, UART, SPI)
12:7:38.787 -> 
12:7:38.787 -> ########################################
12:7:38.787 -> #### Test PASSED
12:7:38.787 -> ########################################
12:7:38.787 -> 
```

fpistm replied:
Hi @pieterconradie 
About analog pins, in fact you defined 0 analog pins as you have removed al the PIN_Ax. Moreover the `analogInputPin[] `defin pins with no analog capabilities (ex: `D32 `which is `PG0`).

pieterconradie replied:
Hi @pieterjanbuntinx,
I think this comment is supposed to be for you.
Kind regards,
Pieter

fpistm replied:
Thank you @pieterconradie  you are right. My mistake 😖

pieterconradie replied:
@fpistm No worries :) Keep up the great work!

pieterjanbuntinx commented:
Hey @fpistm
Thank you for the feedback! 
I think there is some misunderstanding about the ESC part of this board. This not an RC speed controller, this a board with an EtherCAT Slave Controller. Coincidentally, they have the same abbreviation but they are not the same. So I don't think they belong in the same category. 
At the moment we don't have any documentation or schematics (except for the complete PCB schematics). We will certainly make those and provide them on our website. The website [URL](https://www.elecgator.com/products/info/ethercatduino) provided in the repo currently points to a non-existing page where we will provide all of this together with a link to buy the product. We first wanted to have our board supported by the main repo before finishing the rest. 
Concerning the Astyle checks, is there some documentation on how can check them myself? I tried running the Astyle.py script in the CI directory but it complained about some connection it couldn't make.

fpistm commented:
OK. for the ESC as no documentation is available at this time hard to check 😉 
For Astyle, you need to have astyle installed and available in the path or providing the path where to find it. 
https://github.com/stm32duino/wiki/wiki/Astyle

Then you can call it from repo root path like this (specify the path to check instead of checking the full repo):
`python CI/astyle/astyle.py -r variants/STM32F7xx/F745Z(E-G)T_F746Z(E-G)(T-Y)_F750Z8T_F756ZG(T-Y)/`

fpistm commented on variants/STM32F7xx/F745Z(E-G)T_F746Z(E-G)(T-Y)_F750Z8T_F756ZG(T-Y)/variant_ETHERCAT_DUINO.cpp:
```diff
@@ -0,0 +1,156 @@
+*/
+WEAK void SystemClock_Config(void)
+{
+  RCC_OscInitTypeDef RCC_OscInitStruct = {0};
+  RCC_ClkInitTypeDef RCC_ClkInitStruct = {0};
+  RCC_PeriphCLKInitTypeDef PeriphClkInitStruct = {0};
```
```suggestion
  RCC_OscInitTypeDef RCC_OscInitStruct = {};
  RCC_ClkInitTypeDef RCC_ClkInitStruct = {};
  RCC_PeriphCLKInitTypeDef PeriphClkInitStruct = {};
```

fpistm commented on variants/STM32F7xx/F745Z(E-G)T_F746Z(E-G)(T-Y)_F750Z8T_F756ZG(T-Y)/variant_ETHERCAT_DUINO.cpp:
```diff
@@ -0,0 +1,156 @@
+  RCC_OscInitStruct.PLL.PLLP = RCC_PLLP_DIV2;
+  RCC_OscInitStruct.PLL.PLLQ = 9;
+  if (HAL_RCC_OscConfig(&RCC_OscInitStruct) != HAL_OK) {
+  }
```
```suggestion
    Error_Handler();
  }
```

fpistm commented on variants/STM32F7xx/F745Z(E-G)T_F746Z(E-G)(T-Y)_F750Z8T_F756ZG(T-Y)/variant_ETHERCAT_DUINO.cpp:
```diff
@@ -0,0 +1,156 @@
+  /** Activate the Over-Drive mode
+  */
+  if (HAL_PWREx_EnableOverDrive() != HAL_OK) {
+  }
```
```suggestion
    Error_Handler();
  }
```

fpistm commented on variants/STM32F7xx/F745Z(E-G)T_F746Z(E-G)(T-Y)_F750Z8T_F756ZG(T-Y)/variant_ETHERCAT_DUINO.cpp:
```diff
@@ -0,0 +1,156 @@
+  RCC_ClkInitStruct.APB2CLKDivider = RCC_HCLK_DIV2;
+
+  if (HAL_RCC_ClockConfig(&RCC_ClkInitStruct, FLASH_LATENCY_7) != HAL_OK) {
+  }
```
```suggestion
    Error_Handler();
  }
```

fpistm commented on variants/STM32F7xx/F745Z(E-G)T_F746Z(E-G)(T-Y)_F750Z8T_F756ZG(T-Y)/variant_ETHERCAT_DUINO.cpp:
```diff
@@ -0,0 +1,156 @@
+  PeriphClkInitStruct.I2c1ClockSelection = RCC_I2C1CLKSOURCE_PCLK1;
+  PeriphClkInitStruct.Clk48ClockSelection = RCC_CLK48SOURCE_PLL;
+  if (HAL_RCCEx_PeriphCLKConfig(&PeriphClkInitStruct) != HAL_OK) {
+  }
```
```suggestion
    Error_Handler();
  }
```

fpistm commented on variants/STM32F7xx/F745Z(E-G)T_F746Z(E-G)(T-Y)_F750Z8T_F756ZG(T-Y)/variant_ETHERCAT_DUINO.h:
```diff
@@ -0,0 +1,203 @@
+#define PE5                     50
+#define PE6                     49
+#define PE7                     19
+#define PE8                     18
+#define PE9                     30
+#define PE10                    31
+#define PE11                    28
+#define PE15                    29
+#define PF3                     45
+#define PF4                     44
+#define PF10                    43
+#define PF11                    36
+#define PF12                    35
+#define PF13                    34
+#define PF14                    21
+#define PF15                    20
+#define PG0                     32
+#define PG1                     33
+#define PG2                     26
+#define PG3                     27
+#define PG4                     24
+#define PG5                     25
+#define PG6                     22
+#define PG7                     23
+#define PG9                     37
+#define PG10                    38
+#define PG11                    39
+#define PG12                    40
+#define PG13                    41
+#define PG14                    42
+#define PH0                     112
+#define PH1                     113
+#define PB3                     54
+#define PB8                     55
+#define PB9                     56
+#define PE12                    57
+#define PE13                    58
+#define PE14                    59
+#define PG15                    60
+#define PF2                     61
```
Please reorder for better readability.
Moreover some are higher than   #define NUM_DIGITAL_PINS        62
Ex: 
#define PH0                     112
#define PH1                     113

So they should  be removed.

pieterjanbuntinx replied:
Hopefully, everything is fixed now :)

fpistm replied:
Hope so 😉 
Anyway I've only made a quick review 🤣 I will go deeply soon.

fpistm commented on README.md:
```diff
@@ -460,6 +460,12 @@ User can add a STM32 based board following this [wiki](https://github.com/stm32d
+
+| Status | Device(s) | Name | Release | Notes |
+| :----: | :-------: | ---- | :-----: | :---- |
+| :green_heart:  | STM32F746ZET | [EtherCATduino](https://www.elecgator.com/products/info/ethercatduino) | *2.1.0* |  |
```
```suggestion
| :yellow_heart:  | STM32F746ZET | [EtherCATduino](https://www.elecgator.com/products/info/ethercatduino) | **2.1.0** |  |
```

fpistm reviewed (changes requested):
Hi @pieterjanbuntinx 

I made a deeper review and found several issue.
I've made a branch on my fork with
https://github.com/fpistm/Arduino_Core_STM32/tree/pr-1414-ethercatduino-review

* This PR has been rebased and squashed on top of the master
* Fix Analog pin definition
    (only 3 defined in the variant.h have the Ax capabilities)
* Comment DAC as no pin defined
* Add USB pin definition (PA11/PA12)
* Add link to the board section in the README.md
* Reorder boards.txt (alphabetical order for manufacturer)
* Remove node definition
* Add custom peripheral pins as not all pins available and allow to save space
*  All pins not defined in the variant.h has been commented
* Comment some extra pins or HAL module definition
  as not all pins available for the same peripheral instance
 Ex: No QSPI CLK defined or SDMMC CMD or ETH_REF_CLK
* Fixed system core clock config

About analog pin, I saw that it should be 12 so you have to defined in the variant.h all the pins which can be used. Same for ethernet (I guess you need Ethernet?)
About the clock config, you probably test it before adding the `Error_handler()`?
The value used set the freq at 675 MHz because the default `HSE_VALUE `is 25 MHz.
I guess you have a 8MHz ? In that case `HSE_VALUE `have to be defined in the variant_ETHERCAT_DUINO.h

`#define HSE_VALUE             8000000`

My update are based on the variant.h definition. With a schematics or documentation it is easier to properly review it.

pieterjanbuntinx commented:
> Hi @pieterjanbuntinx
> 
> I made a deeper review and found several issue.
> I've made a branch on my fork:
> https://github.com/fpistm/Arduino_Core_STM32/tree/pr-1414-ethercatduino-review
> 
> * This PR has been rebased and squashed on top of the master
> * Fix Analog pin definition
>   (only 3 defined in the variant.h have the Ax capabilities)
> * Comment DAC as no pin defined
> * Add USB pin definition (PA11/PA12)
> * Add link to the board section in the README.md
> * Reorder boards.txt (alphabetical order for manufacturer)
> * Remove node definition
> * Add custom peripheral pins as not all pins available and allow to save space
> * All pins not defined in the variant.h has been commented
> * Comment some extra pins or HAL module definition
>   as not all pins available for the same peripheral instance
>   Ex: No QSPI CLK defined or SDMMC CMD or ETH_REF_CLK
> * Fixed system core clock config
> 
> About analog pin, I saw that it should be 12 so you have to defined in the variant.h all the pins which can be used. Same for ethernet (I guess you need Ethernet?)
> About the clock config, you probably test it before adding the `Error_handler()`?
> The value used set the freq at 675 MHz because the default `HSE_VALUE `is 25 MHz.
> I guess you have a 8MHz ? In that case `HSE_VALUE `have to be defined in the variant_ETHERCAT_DUINO.h
> 
> `#define HSE_VALUE 8000000`
> 
> My update are based on the variant.h definition. With a schematics or documentation it is easier to properly review it.

pieterjanbuntinx commented:
Hi @fpistm,

Thank you for the review! (please ignore the accidental closing and reopening of this PR, I'm kind of a Github noob)
I've merged the modifications in your fork with mine. 

- I now have fixed the analog Pin definitions. They now should be correctly defined and they all work physically (I've tested them).
- I've added pin definitions for DAC because they are physically present on the board (I really have to add a pin definition to the manual)
- USB pins added
- readme merged
- boards.txt merged
- removed unused HAL modules (QSPI and ETHERNET are used in the library so these HAL modules have to be defined in the library itself)
- merged the clock config (changes in clock config for QSPI, ETHERNET and I2C have to be made in the library)

fpistm commented:
Thanks @pieterjanbuntinx 
About the clock config what is the value of your HSE ? About QSPI, ETHERNET and I2C  clock config you should have nothing to do if you used library. Anyway if you manage them in your library then yes you will have to handle them.
I will fetch again the PR and review it.

pieterjanbuntinx commented:
@fpistm , the HSE value is 8 MHz. I see I then maybe have merged a little too much from your modifications. I'm going to test all of this thoroughly with my board and I'll get back to you on this.

fpistm commented:
> 
> 
> @fpistm , the HSE value is 8 MHz. I see I then may have merged a little too much from your modifications. I'm going to test all of this thoroughly with my board and I'll get back to you on this.

So you have to add this in the variant header:
`#define HSE_VALUE 8000000`

pieterjanbuntinx commented:
Hi @fpistm, 

How can I properly override the SystemClock_Config() function in my library? I tried copy+pasting my initial systemclock_config function to the library and calling it in the init function but that doesn't seem to apply everything. I have to change these two:
```
  RCC_OscInitStruct.PLL.PLLM = 25;
  RCC_OscInitStruct.PLL.PLLN = 432;
```
in SystemClock_config in the variant.cpp file back to respectively 4 and 216 to make the USB serial work again. I also have to define the following in the variant.h file to get my QSPI for the EtherCAT chip working:
```
#if !defined(HAL_QSPI_MODULE_DISABLED)
  #define HAL_QSPI_MODULE_ENABLED
#endif
```
Copying the same lines into my main library file doesn't seem to work. In hal_conf.custom.h `#define HAL_QSPI_MODULE_ENABLED` is defined, I would think that would be enough for everything to work.

fpistm commented:
@pieterjanbuntinx 
as the system clock config relies to your variant you can simply do what you need in this one. To redefine the clock config if the file is a cpp file then you need to add `extern "C"` before.
https://github.com/stm32duino/wiki/wiki/Custom-definitions#systemclock_config

Same for  the QSPI enable it by default in the variant. I don't think the `hal_conf_custom.h` works if you add it in the library as library is build after the core moreover adding it in the library would prevent other user to add it. If you really need it then it should be in sketch example like this a user can complet it and the core will use it.
https://github.com/stm32duino/wiki/wiki/HAL-configuration#customize-hal-or-variant-definition

I will review this PR and add proper config.

pieterjanbuntinx commented:
@fpistm 
I already was using the `extern "C"` but omitting the `WEAK` keyword seemed to make it work now. 
`hal_conf.custom.h` wasn't needed after all so I removed it from my library. 
I've committed the changes in the variant.h file.

fpistm commented:
What are the pins used by the QSPI and the Ethernet? Seems they are not defined?
Do you used SD?

pieterjanbuntinx commented:
Ethernet of the STM32 chip is not used, the Ethernet jacks are driven by an external controller that is interfaced using QSPI. The QSPI pins are configured internally in the library using the STM32 method instead of Arduino's. No, we don't use SD

fpistm commented:
> 
> 
> Ethernet of the STM32 chip is not used, the Ethernet jacks are driven by an external controller that is interfaced using QSPI. The QSPI pins are configured internally in the library using the STM32 method instead of Arduino's. No, we don't use SD

OK it is clear, anyway some SQPI pins are not defined, mainly PB2 for the clock so probably other one are not defined.
I've almost finished to clean up the PR it misses only those information.

pieterjanbuntinx commented:
I'll add all the pin definitions used to communicate with EtherCAT chip

fpistm commented:
OK.
I've updated my branch with all the fix (including the correct clock config with an HSE at 8MHz
https://github.com/fpistm/Arduino_Core_STM32/tree/pr-1414-ethercatduino-review

pieterjanbuntinx commented:
I have now added all the used pins definitions of the EtherCAT controller. I have also reverted the PLLM and PLLN values. Otherwise USB virtual serial won't work.

fpistm commented:
Thanks @pieterjanbuntinx 
I've updated my branch with the new pins.
Anyway your branch is not complet you don't have the `PeripheralPins_ETHERCAT_DUINO.c`
I've also made some formatting and other minor update. That should be fine you replace your branch by the mine which is also squashed in only one commit.

fpistm commented:
@pieterjanbuntinx  I also can push my branch on your fork and replace your branch if it is more simple for you?
Just a tip for future PR. Avoid to use the master/main branch this ease to keep your fork synced with the original repo and rebase.

pieterjanbuntinx commented:
Hi @fpistm 
I've already merged your changes but that could be easier in the future :) 
I think now all should be fine except perhaps the ALT pins definitions.

fpistm commented on .github/ISSUE_TEMPLATE/bug_report.md:
```diff
@@ -1,8 +1,8 @@
 ---
-name: "\U0001F41B Bug report"
```
This file should not be modified

fpistm commented on .github/ISSUE_TEMPLATE/config.yml:
```diff
@@ -1,8 +0,0 @@
-blank_issues_enabled: false
```
dito

fpistm commented on .github/ISSUE_TEMPLATE/feature-request-improvement.md:
```diff
@@ -1,26 +1,22 @@
 ---
-name: "⭐ Feature request/improvement"
```
dito

fpistm commented on .github/ISSUE_TEMPLATE/new-variant-request.md:
```diff
@@ -1,5 +1,5 @@
 ---
-name: "🎁 New variant request"
```
dito

fpistm commented on libraries/SrcWrapper/src/stm32/pinmap.c:
```diff
@@ -68,64 +68,35 @@ const uint32_t pin_map_ll[16] = {
  * Configure Analog dualpad switch if necessary
- * LL_AnalogSwitch: LL define to be used to configure Analog switch
  */
-static void configure_dualpad_switch(PinName pin, int function, uint32_t LL_AnalogSwitch)
```
dito

fpistm commented on variants/STM32F7xx/F745Z(E-G)T_F746Z(E-G)(T-Y)_F750Z8T_F756ZG(T-Y)/variant_ETHERCAT_DUINO.cpp:
```diff
@@ -0,0 +1,188 @@
+  PA_7,     // A9
+  PC_4,     // A10
+  PC_5,     // A11
+  PD_0,     // CANRX
+  PD_1,     // CANTX
+  PA_5,     // DAC0
+  PA_4,     // DAC1
```
```suggestion
  PA_5,     // DAC0
  PA_4,     // DAC1
  PD_0,     // CANRX
  PD_1,     // CANTX
```

fpistm reviewed (changes requested):
Your merge has some side effect, you have reverted last commits of the original repo (as your master was not synced).

pieterjanbuntinx commented:
Hi @fpistm, 
I've tried some things to fix my bad merge attempt but I don't see how I can fix them without manually editing. Can you help out :) ?

fpistm commented:
Hi @pieterjanbuntinx 
You can fetch your fork, all is solved.

fpistm reviewed (approved):
LGTM

pieterjanbuntinx commented:
Great! Thanks for approving and all your help!

fpistm commented:
FYI, when you create a PR by default you allow maintainers to edit. This means you allow me to push on your master branch as you used this branch for the PR. That's why I always advice to create a dedicated branch for PR to avoid master edition on your fork.

fpistm commented:
> 
> 
> Great! Thanks for approving and all your help!

Welcome, anyway it would be good you do a final test to avoid any issue. Rerun check variant and stuff you consider useful to validate your board (USB, Serial, I2C, SPI,...)

When done let me know if all is OK then we could merge.

pieterjanbuntinx commented:
> FYI, when you create a PR by default you allow maintainers to edit. This means you allow me to push on your master branch as you used this branch for the PR. That's why I always advice to create a dedicated branch for PR to avoid master edition on your fork.

Okay, I'll try to think of that next time I do a PR. I've given you access now.

> > Great! Thanks for approving and all your help!
> 
> Welcome, anyway it would be good you do a final tests session to avoid any issue. Rerun check variant and stuff you consider useful to validate your board (USB, Serial, I2C, SPI,...)
> 
> When done let me know if all is OK then we could merge.

At the moment I'm busy doing that, I will let you know if everything works.

fpistm commented:
> Okay, I'll try to think of that next time I do a PR. I've given you access now.

Thanks bu no need to invite me. It is a feature of the PR mechanism :wink:
https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork

fpistm commented:
Hi @pieterjanbuntinx 

> At the moment I'm busy doing that, I will let you know if everything works.

Is this OK?

pieterjanbuntinx commented:
Hey @fpistm 
I'm currently having some problems gettings all peripherals to work. For example, `HardwareSerial serial(PC10, PC11)` (UART4 or 3) works perfectly but doing the same for `PE7` and `PE8` (UART7) or the other UARTs don't seem to work. The same for I2C, I2C1 on `PB8` and `PB9` works perfectly but the other I2C busses won't work. I also wanted to add support for the remaining Analog pins, the DAC0 and DAC1 pins (PA5, PA4) and D44 and D45 (PF4, PF3), but those also won't work. I was thinking this all has to do with Alternate Functions but those seem to be correctly set in the PeripheralPins_*.c file. I tried using for example `HardwareSerial serial((PE8 | ALT8),(PE7 | ALT8)` but that doesn't seem to solve it. Am I still missing something?

All pins work perfectly as digital input or output and A0 to A12 also work perfectly as Analog inputs. I've attached a pin diagram I have created for our board.

![EtherCATduino_pinout-01](https://user-images.githubusercontent.com/34648386/122760610-99618100-d29b-11eb-942e-4c286e678de0.png)

fpistm commented:
> I tried using for example `HardwareSerial serial((PE8 | ALT8),(PE7 | ALT8)`

There is no link between AF (Alternate Function) and the ALTx pins which stands for alternate pins for a peripheral.
So simply use `HardwareSerial serial(PE8, PE7);`

For the ALTx usage, an example is if you want to use `USART3 `with  `PC10` and `PC11` then you have to do:
`HardwareSerial serial(PC11_ALT1, PC10_ALT1);`


About the DAC it is managed by default when you used the `analogWrite()` API with those pins.

pieterjanbuntinx commented:
Hey @fpistm, 

> There is no link between AF (Alternate Function) and the ALTx pins which stands for alternate pins for a peripheral.
So simply use HardwareSerial serial(PE8, PE7);

That's the way I started doing it. I was just experimenting to try and get it to work. Currently the following works correctly: I2C4 (PD12 PD13), I2C1 (PB8 PB9), UART6 (PC6 PC7), UART6 (PG9, PG14), UART4 (PC10, PC11), UART3 (PC10, PC11), UART2 (PA2, PA3), UART4 (PA0, PA1), all analog pins except the 4 listed below, all digital inputs/outputs.
The following won't work: I2c4 (PF14, PF15), I2c2 (PB11, PB10), I2C3 (PA8, PC9), UART1 (only tx broken out -> PA9), UART7 (PE7, PE8), UART3 (PB11, PB10), UART2 (PD6, PD5), both DAC outputs (PA5, PA4) as DAC or as ADC and PF4 and PF3 used as ADC.

I've tried this on multiple boards so hardware failure seems unlikely. I've done most testing on PlatformIO but I've now also tried most of the testing in the Arduino IDE by flashing over DFU.

fpistm commented:
@pieterjanbuntinx 
Why have you merged branch... you add again commits in this PR not related to it...

About your issue with the pins I have no clue for you. This is generic and know to work.

pieterjanbuntinx commented:
@fpistm sorry, I shouldn't have done that. I've seen you have reverted it. 

I'll try to see with the debugger if the registers get correctly set. Have you had anything similar to this before?

fpistm commented:
Don't know how you managed your branch but you loose some commit from the master.
I've realigned it properly but this is the last time. So please pay attention.

fpistm commented:
I've fix the astyle issue.

> I'll try to see with the debugger if the registers get correctly set. Have you had anything similar to this before?

No, I've tested the Nucleo F746 and this is fine.

fpistm commented:
Just to confirm, I've retest with the Nucleo F746ZG:
* I2c4 (PF14, PF15)
* UART3 (PB11, PB10


About this: UART1 (only tx broken out -> PA9)
You can configure the Serial to be half duplex so only Tx pin are required --> https://github.com/stm32duino/wiki/wiki/API#enable-half-duplex-mode

pieterjanbuntinx commented:
Hey @fpistm,
Thank you for retesting! I have just found out I have made a stupid mistake. The order of one header of pins was drawn inverted on the schematics relative to the PCB design. This explains some of the non-working peripherals. I have also found out I have a Nucleo F746ZG board over here so I can now check everything out on there when in doubt. I'll try your suggestion for UART1.  
Sorry for the trouble. I will post a new comment here when everything is fixed.

pieterjanbuntinx commented:
Hey @fpistm,
Thanks for testing! I have now solved all issues I was having. I made the stupid mistake to look at the wrong pinout. I also found out we have a Nucleo F746 so I was able to double-check it on there as well.

fpistm commented:
@pieterjanbuntinx 
Fine!
So is this PR now OK on your side ?

pieterjanbuntinx commented:
Yes, for me it's ready to be merged :)

fpistm commented on variants/STM32F7xx/F745Z(E-G)T_F746Z(E-G)(T-Y)_F750Z8T_F756ZG(T-Y)/variant_ETHERCAT_DUINO.h:
```diff
@@ -0,0 +1,303 @@
+// Serial pins
+#define TX0                     PC10          // UART4      
+#define RX0                     PC11          // UART4     
+#define TX0_ALT                 PC10_ALT1     // UART3 
+#define RX0_ALT                 PC11_ALT1     // UART3
```
```suggestion
#define TX0_ALT                 PC10_ALT1     // USART3 
#define RX0_ALT                 PC11_ALT1     // USART3
```

fpistm commented on variants/STM32F7xx/F745Z(E-G)T_F746Z(E-G)(T-Y)_F750Z8T_F756ZG(T-Y)/variant_ETHERCAT_DUINO.h:
```diff
@@ -0,0 +1,303 @@
+#define RX3                     PD6           // UART2
+
+// I2C pins
+#define SDA0                    PF15          // I2C4_ALT
+#define SCL0                    PF14          // I2C4_ALT
```
Why ALT ?

```suggestion
#define SDA0                    PF15          // I2C4
#define SCL0                    PF14          // I2C4
```

pieterjanbuntinx replied:
ALT was something I called it in my testing firmware, so that is no longer relevant

fpistm commented on variants/STM32F7xx/F745Z(E-G)T_F746Z(E-G)(T-Y)_F750Z8T_F756ZG(T-Y)/variant_ETHERCAT_DUINO.h:
```diff
@@ -0,0 +1,303 @@
+#define TX2                     PB10          // UART3
+#define RX2                     PB11          // UART3
+
+#define TX3                     PD5           // UART2
+#define RX3                     PD6           // UART2
```
```suggestion
#define TX3                     PD5           // USART2
#define RX3                     PD6           // USART2
```

fpistm commented on variants/STM32F7xx/F745Z(E-G)T_F746Z(E-G)(T-Y)_F750Z8T_F756ZG(T-Y)/variant_ETHERCAT_DUINO.h:
```diff
@@ -0,0 +1,303 @@
+#define TX1                     PE8           // UART7
+#define RX1                     PE7           // UART7
+
+#define TX2                     PB10          // UART3
+#define RX2                     PB11          // UART3
```
```suggestion
#define TX2                     PB10          // USART3
#define RX2                     PB11          // USART3
```

fpistm reviewed (approved):
LGTM
